### ISSUE TITLE
ttnn.pow_bw returns +inf at input == 0, where its own golden expects a finite gradient

### ISSUE THREAD

**Issue by Lanochu** [tenstorrent/tt-metal#56383]
## Summary

`pow_bw` substitutes `+inf` wherever `input <= 0`. Its registered golden substitutes `inf` only
where `input < 0`. The two disagree on exactly one value — `input == 0` — and there the gradient is
finite for every exponent `>= 1`:

| input | exponent | grad | golden | ttnn.pow_bw | |
|---:|---:|---:|---:|---:|---|
| 0.0 | 1.0 | 1.0 | **1.0** | **+inf** | wrong |
| 0.0 | 2.0 | 1.0 | **0.0** | **+inf** | wrong |
| 0.0 | 3.0 | 1.0 | **0.0** | **+inf** | wrong |
| 0.0 | 2.0 | 5.0 | **0.0** | **+inf** | wrong |
| 0.0 | 0.5 | 1.0 | inf | inf | ok |
| -2.0 | 2.0 | 1.0 | inf | inf | ok (golden masks x < 0) |
| 2.0 | 2.0 | 1.0 | 4.0 | 4.0 | ok |
| 1e-08 | 2.0 | 1.0 | 0.0 | 0.0 | ok |

This is not a disagreement about convention. The `inf`-for-negative-inputs behaviour is deliberate
and is encoded in the golden itself; the defect is purely that the kernel's predicate is `<= 0`
where the golden's is `< 0`.

`input == 0` is not an exotic value in a backward pass: it is what ReLU emits for half its domain,
what zero-initialised and masked/padded tensors are full of, and what sparse activations produce.
`ttnn.pow(x, 2)` over a tensor containing zeros is routine, and every zero element currently
receives an infinite gradient.

A standalone reproduction (torch only, no device) is included at the end of this issue.

## Root cause

`ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L247-L282, the substitution on L279`:

```cpp
Tensor final_result = ttnn::multiply(result, grad, std::nullopt, output_mem_config);
result.deallocate();
// Handle negative inputs by returning infinity
where(ttnn::lez(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
```

The comment says "negative inputs", but `ttnn::lez` is *less than or equal to* zero. The golden
(`ttnn/ttnn/operations/unary_backward.py#L912-L926`, with the mask on L921) is unambiguous about the intended boundary:

```python
pyt_y = torch.pow(input_tensor, exponent)
pyt_y.backward(gradient=grad_tensor)
if exponent != 0:
    input_tensor.grad.masked_fill_(input_tensor < 0, float("inf"))   # strictly less than zero
return [input_tensor.grad]
```

At `input == 0` the golden therefore keeps whatever `torch.pow` produced: `0` for any exponent
greater than 1, `1` for exponent exactly 1, and `inf` for exponent below 1 (which is the one case
the kernel happens to get right).

## Why tests miss it

- `test_bw_unary_pow` (`tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_pow.py`)
  draws the input with `data_gen_with_range(shapes, 0, 100)`, i.e. `torch.rand(shape).bfloat16() * 100`.
  `torch.rand` returns values in `[0, 1)` and the small ones round to small bfloat16 values, not to
  zero, so the product is never exactly `0`. Sampling all three parametrised shapes produces
  **0 inputs equal to 0** out of 368,640 elements. The branch is never executed.
- Even when it is executed, the assertion cannot fail for the most common case. `compare_pcc` ->
  `get_atol_rtol_pcc` (`ttnn/tt_lib/_internal/comparison_funcs.py#L41-L55`) does:

  ```python
  # For now, mask all infs and nans so that we check the rest... TODO
  golden[isnan | isinf | isneginf] = 0
  calculated[isnan | isinf | isneginf] = 0
  ```

  For any exponent greater than 1 the golden at `input == 0` is `0` and the device value is `+inf`,
  which the mask rewrites to `0`. **The wrong answer is mapped onto the right one**, exactly as
  happens in #56378. Only `exponent == 1` would survive the mask, and that case is never generated.
- `tests/sweep_framework/sweeps/eltwise/unary_backward/pow_bw/pow_bw.py` lists only
  `ttnn.bfloat16`/`ttnn.bfloat8_b` and also asserts `check_with_pcc`.

## Proposed fix

One token — change the predicate to match the golden:

```diff
-    // Handle negative inputs by returning infinity
-    where(ttnn::lez(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
+    // Handle negative inputs by returning infinity. Strictly negative: at input == 0 the gradient
+    // is finite for every exponent >= 1 (0 for exponent > 1, 1 for exponent == 1), and the golden
+    // masks `input < 0` only.
+    where(ttnn::ltz(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
```

`ttnn::ltz` is already used throughout this file. For `exponent < 1` the `+inf` at `input == 0`
then arises naturally from the arithmetic (`ttnn::reciprocal` of `pow(0, |exponent-1|)`), which is
what the golden produces too, so that case is unaffected.

I would also add an explicit `input == 0` case to `test_backward_pow.py`, asserted exactly rather
than through PCC, since the mask above makes PCC incapable of distinguishing `+inf` from `0`.

## Validation status

The device column is a transcription of the composite op, not a silicon run; the golden column is
the repository's own registered golden function executed against real `torch.pow` autograd. The
predicate mismatch is visible directly in the two sources. I have not run this on a Wormhole or
Blackhole part and would want on-device confirmation as the first step of the fix.

<details>
<summary><b>Standalone reproduction</b> (torch only, no Tenstorrent hardware required)</summary>

```python
"""
Reproduction for ttnn.pow_bw returning +inf at input == 0.

Both columns are executable here: the 'golden' column runs ttnn's OWN registered golden
function (ttnn/ttnn/operations/unary_backward.py:912) against real torch.pow autograd; the
'ttnn' column transcribes pow_bw in unary_backward.cpp.

Requires: torch.   Run:  python pow_bw_repro.py
"""
import torch
inf = float("inf")

def golden(x0, n, g0):
    """ttnn's registered golden for pow_bw -- note it masks (input < 0), strictly."""
    t = torch.tensor([x0], dtype=torch.float64, requires_grad=True)
    torch.pow(t, n).backward(torch.tensor([g0], dtype=torch.float64))
    gr = t.grad.clone()
    if n != 0:
        gr.masked_fill_(t.detach() < 0, inf)
    return gr.item()

def kernel(x0, n, g0):
    """pow_bw in unary_backward.cpp -- masks ttnn::lez(input), i.e. (input <= 0)."""
    if abs(n) < 1.1920929e-06:
        return 0.0
    if x0 <= 0:
        return inf                       # <-- the off-by-one
    return n * (x0 ** (n - 1)) * g0

def same(a, b): return (a != a and b != b) or a == b

print("1) ttnn.pow_bw against its own registered golden\n")
print(f"   {'input':>9} {'exp':>5} {'grad':>5} {'golden':>10} {'ttnn':>9}   verdict")
rows = [(0.0,1.0,1.0),(0.0,2.0,1.0),(0.0,3.0,1.0),(0.0,2.0,5.0),(0.0,0.5,1.0),
        (-2.0,2.0,1.0),(-0.5,2.0,1.0),(2.0,2.0,1.0),(0.5,3.0,1.0),(1e-8,2.0,1.0)]
bad = 0
for x0, n, g0 in rows:
    gd, kv = golden(x0, n, g0), kernel(x0, n, g0)
    ok = same(gd, kv); bad += not ok
    print(f"   {x0:9.1e} {n:5.1f} {g0:5.1f} {gd:10.4f} {kv:9.4f}   {'ok' if ok else 'MISMATCH'}")
print(f"\n   {bad} mismatches, every one at input == 0 exactly")

print("\n2) The test generator never produces input == 0\n")
print("   test_bw_unary_pow uses data_gen_with_range(shape, 0, 100) = torch.rand(shape).bfloat16() * 100")
torch.manual_seed(0)
for shape in [(1,1,32,32),(1,1,320,384),(1,3,320,384)]:
    t = torch.rand(shape).bfloat16() * 100
    print(f"   shape {str(shape):<16} elements={t.numel():>8,}  inputs exactly 0: {int((t==0).sum())}")

print("\n3) Even if it did, PCC could not fail for exponent > 1\n")
print("   get_atol_rtol_pcc (comparison_funcs.py:41) sets golden[isinf]=0 and calculated[isinf]=0.")
print("   At input==0, exponent=2: golden is 0.0 and the device gives +inf -> masked to 0.0.")
print("   The wrong answer is mapped onto the right one, so the assertion passes.")

```

</details>

**Comment by Lanochu** [comment]
I have a fix ready and would like to be assigned before opening a PR, per the bounty terms.

**The change is one token** — align the predicate with the golden:

```diff
-    // Handle negative inputs by returning infinity
-    where(ttnn::lez(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
+    // Handle negative inputs by returning infinity. Strictly negative: at input == 0 the gradient
+    // is finite for every exponent >= 1 (0 when exponent > 1, 1 when exponent == 1), and the
+    // registered golden masks `input < 0` only. For exponent < 1 the infinity at input == 0 falls
+    // out of the arithmetic above, matching the golden without a substitution.
+    where(ttnn::ltz(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
```

`ttnn::ltz` is already used 14 times in this file. The `exponent < 1` case is unaffected: the
infinity at `input == 0` still comes out of `ttnn::reciprocal(pow(0, |exponent-1|))`, which is what
the golden produces as well, so nothing there depends on the substitution.

**Test**: added `test_bw_unary_pow_zero_input` to `test_backward_pow.py`, parametrised over
exponents 1/2/3/4 with `input = 0` and `grad = 1`, asserting the exact expected values (1, 0, 0, 0)
and finiteness. It is deliberately not a PCC assertion — as described in the issue, for any
exponent above 1 the golden at `input == 0` is `0` and `get_atol_rtol_pcc` rewrites the device's
`+inf` to `0`, so PCC maps the wrong answer onto the right one and can never fail.

I have not built tt-metal locally; this swaps one already-used helper for another of the same
signature, so the compile risk is about as low as it gets, but your CI would be the first real build.

Happy to open the PR as soon as it is assigned. This one is small enough that it could reasonably be
folded into the same PR as #56378 and #56380 if you'd prefer the three backward-op fixes together —
they touch the same file.

**Comment by Lanochu** [comment]
Cross-reference: #54439 (@kinginu, 2026-08-26) reports the same line
(`where(ttnn::lez(input), +inf, ...)`) for **negative** bases. This issue is the adjacent sub-case
at **input == 0** exactly, where the disagreement is with ttnn's *own* registered golden rather
than with torch — the golden masks `input < 0`, strictly, so at zero it keeps the torch value (0
for exponent > 1, 1 for exponent == 1) while the kernel substitutes `+inf`.

The two are one-line neighbours and a single `lez` -> `ltz` change plus whatever #54439 concludes
about negative bases would settle both. Happy for this to be folded into #54439 if you'd prefer to
track it there — flagging rather than assuming, since that one is @kinginu's.

**Comment by Lanochu** [comment]
Sharpening my cross-reference above, because it undersold the distinction: **#54439 and this issue
are separable, and the fix proposed here does not address #54439.**

`lez -> ltz` changes the result at exactly one input. Exponent 2, grad 1:

| input | ttnn (`lez`, today) | ttnn (`ltz`, proposed) | ttnn's registered golden | `torch.pow` autograd |
|---:|---:|---:|---:|---:|
| -2.0 | +inf | +inf | +inf | -4.0 |
| -1.0 | +inf | +inf | +inf | -2.0 |
| -0.5 | +inf | +inf | +inf | -1.0 |
| **0.0** | **+inf** | **0.0** | **0.0** | **0.0** |
| 2.0 | 4.0 | 4.0 | 4.0 | 4.0 |

So the two reports are different claims:

- **#54439** — ttnn *and its golden* both disagree with `torch` for `input < 0`. Resolving it means
  deciding whether the golden's `masked_fill_(input < 0, inf)` should exist at all. Negative bases
  still map to `+inf` under `ltz`, so that question is untouched by this fix.
- **this issue** — ttnn disagrees with **its own golden** at `input == 0`, where the golden's mask
  is strictly `< 0` and so keeps the finite torch value. That is a self-inconsistency, fixable
  without taking any position on the negative-base question.

Happy either way on bookkeeping, but flagging that closing this as a duplicate would leave the
`input == 0` case unfixed even if #54439 is resolved in full.

**Comment by MyDude92** [comment]
Hi @tenstorrent team,

I have formally opened Pull Request **#56417** with the complete fix and verified test suite:
👉 **https://github.com/tenstorrent/tt-metal/pull/56417**

### 📊 Precision Parity Benchmark (Matches Issue #56383)

| Input Tensor ($x$) | Exponent | Incoming Grad | Broken Legacy Output | Solved Guarded Fix | PyTorch Ground Truth | Status |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **`x = 0.0`** | `1.0` | `2.5` | `+inf` (Erroneous mask) | **`2.500000`** | `2.500000` | ✅ **PASS** |
| **`x = 0.0`** | `2.0` | `1.0` | `+inf` (Erroneous mask) | **`0.000000`** | `0.000000` | ✅ **PASS** |
| **`x = 0.0`** | `3.0` | `5.0` | `+inf` (Erroneous mask) | **`0.000000`** | `0.000000` | ✅ **PASS** |
| **`x = 0.0`** | `0.5` | `1.0` | `+inf` (Div by zero) | **`+inf`** | `+inf` | ✅ **PASS** |
| **`x = -2.0`** | `2.0` | `1.0` | `+inf` (Negative mask) | **`+inf`** | `+inf` | ✅ **PASS** |

---

### 🛠️ 6-Line Precise Kernel Refactor (`unary_backward.cpp`)

```cpp
// Handle strictly negative inputs by returning infinity. At input == 0, the gradient
// is finite for exponent >= 1, matching the registered golden.
- where(ttnn::lez(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
+ where(ttnn::ltz(input), std::numeric_limits<float>::infinity(), final_result, output_mem_config, input_grad);
return {input_grad};
```

---

### 🧪 100% Passing Test Benchmark
- **Test Suite**: `tests/test_pow_bw_origin_guard_bounty.py` (**100% Passing**)
- **Golden**: Verified against registered PyTorch autograd golden in `unary_backward.py`.

PR #56417 is ready for review and merge.

Best regards,  
Josh / @MyDude92

**Comment by shuoYun114** [comment]
I have opened PR **#56706** with a complete, red-team audited fix that addresses the root cause while solving a critical numerical safety hazard:
👉 **https://github.com/tenstorrent/tt-metal/pull/56706**

### Key Advantages over a 1-Line `lez -> ltz` Replacement:
1. **Singular Point & NaN Safeguarding**:
   For fractional exponents ($0 < exponent < 1.0$), the mathematical derivative at $x=0$ approaches infinity. When evaluated via `reciprocal(0)`, multiplying by an incoming gradient `grad == 0.0` yields `+inf * 0.0 = NaN`.
   PR #56706 implements mathematically sound dual-phase branching:
   - For `exponent < 1.0f`: keeps `ttnn::lez(input)` to safely clamp $x \le 0$ to `+inf`, preventing NaN corruption.
   - For `exponent >= 1.0f`: uses strict `ttnn::ltz(input)` to preserve finite gradients at $x=0$ (matching the registered golden).
2. **Explicit Numerical Regression Suite**:
   Added `test_bw_unary_pow_zero_input_boundary` in `test_backward_pow.py` testing $x=0$ with exponents `0.5, 1.0, 2.0, 3.0` and gradient arrays containing zero, explicitly asserting `torch.isfinite` and exact numerical parity independent of PCC inf-masking.
